File: tests/unit/pybind/actuators/test_MTQ.py
import sys
from pathlib import Path
import numpy as np
import pytest

ROOT = Path(__file__).resolve().parents[4]
sys.path.insert(0, str(ROOT / "build"))

import saltro_py


def valid_axis():
    return np.array([1.0, 0.0, 0.0])


def valid_base_state():
    x = np.zeros(7)
    x[3] = 1.0
    return x


def valid_magnetic_field():
    return np.array([0.0, 1.0e-5, 0.0])


def test_mtq_constructor():
    axis = valid_axis()
    max_dipole = 0.2

    mtq = saltro_py.MTQ(axis, max_dipole)
    assert np.allclose(mtq.axis, axis)
    assert mtq.u_max == max_dipole


def test_mtq_torque_computation():
    axis = np.array([1.0, 0.0, 0.0])
    mtq = saltro_py.MTQ(axis, 0.2)

    x = valid_base_state()
    B_body = np.array([0.0, 1.0e-5, 0.0])
    u = 0.1

    tau = mtq.torque(u, x, B_body)

    expected = np.array([0.0, 0.0, 1.0e-6])
    
    assert np.allclose(tau, expected, atol=1e-12)


def test_mtq_torque_with_different_configurations():
    axis = np.array([0.0, 1.0, 0.0])
    mtq = saltro_py.MTQ(axis, 1.0)

    x = valid_base_state()
    B = np.array([2.0e-5, 0.0, 0.0])
    
    tau = mtq.torque(0.5, x, B)
    
    expected = np.array([0.0, 0.0, -1.0e-5])
    
    assert np.allclose(tau, expected, atol=1e-12)


def test_mtq_torque_is_zero_when_b_parallel_to_axis():
    axis = np.array([1.0, 0.0, 0.0])
    mtq = saltro_py.MTQ(axis, 1.0)

    x = valid_base_state()
    B = np.array([1.0e-5, 0.0, 0.0])

    tau = mtq.torque(1.0, x, B)
    
    assert np.linalg.norm(tau) < 1e-15


def test_mtq_torque_is_zero_when_u_is_zero():
    axis = valid_axis()
    mtq = saltro_py.MTQ(axis, 1.0)

    x = valid_base_state()
    B = valid_magnetic_field()

    tau = mtq.torque(0.0, x, B)
    
    assert np.allclose(tau, 0.0)


def test_mtq_dtorq_du():
    axis = np.array([1.0, 0.0, 0.0])
    mtq = saltro_py.MTQ(axis, 1.0)

    x = valid_base_state()
    B = np.array([0.0, 1.0e-5, 0.0])

    J = mtq.dtorq_du(0.5, x, B)

    B_cross_axis = np.cross(B, axis)
    expected = -B_cross_axis.reshape(1, 3)
    
    assert np.allclose(J, expected, atol=1e-12)


def test_mtq_dtorq_dbasestate_with_zero_dB_dq():
    axis = valid_axis()
    mtq = saltro_py.MTQ(axis, 1.0)

    x = valid_base_state()
    B = valid_magnetic_field()
    dB_dq = np.zeros((4, 3))

    J = mtq.dtorq_dbasestate(0.5, x, B, dB_dq)

    assert np.allclose(J, 0.0)


def test_mtq_ddtorq_dudbasestate_with_non_zero_dB_dq():
    axis = valid_axis()
    mtq = saltro_py.MTQ(axis, 1.0)

    x = valid_base_state()
    B = valid_magnetic_field()

    dB_dq = np.ones((4, 3))

    H = mtq.ddtorq_dudbasestate(0.5, x, B, dB_dq)

    # Tensor3<1,7,3> -> numpy (slices, rows, cols) = (3, 1, 7): one control row,
    # 7 base-state columns, 3 torque-component slices.
    assert H.shape == (3, 1, 7)
    assert np.all(np.isfinite(H))


def test_mtq_torque_matches_u_times_axis_cross_b_for_random_configs():
    """τ_MTQ = u * (axis_hat × B). Validate across random axes and fields
    to catch sign / argument-order regressions that two hand-picked
    cases (x-axis, y-axis) miss."""
    rng = np.random.default_rng(20260514)
    for _ in range(20):
        raw_axis = rng.normal(size=3)
        if np.linalg.norm(raw_axis) < 1e-6:
            continue
        axis_hat = raw_axis / np.linalg.norm(raw_axis)

        mtq = saltro_py.MTQ(raw_axis, 1.0)
        x = valid_base_state()
        B = rng.normal(size=3) * 1e-5
        u = rng.uniform(-0.5, 0.5)

        tau = mtq.torque(u, x, B)
        expected = u * np.cross(axis_hat, B)
        assert np.allclose(tau, expected, atol=1e-15)


def test_mtq_torque_is_perpendicular_to_axis_and_to_b():
    """Cross product → τ ⊥ axis and τ ⊥ B by construction."""
    axis = np.array([0.3, 0.4, 0.866])
    axis_n = axis / np.linalg.norm(axis)
    mtq = saltro_py.MTQ(axis, 1.0)
    x = valid_base_state()
    B = np.array([2e-5, -1e-5, 1.5e-5])

    tau = mtq.torque(0.3, x, B)

    assert abs(np.dot(tau, axis_n)) < 1e-15
    assert abs(np.dot(tau, B)) < 1e-15

File: tests/unit/pybind/actuators/test_MTQ.cpp
#include <catch2/catch_test_macros.hpp>
#include <Eigen/Dense>
#include <saltro/pybind/actuators/MTQ.h>
#include <cmath>

using Vec3 = Eigen::Vector3d;
using BaseState = Eigen::Matrix<double, 7, 1>;

static Vec3 valid_axis()
{
    Vec3 axis;
    axis << 1.0, 0.0, 0.0;
    return axis;
}

static BaseState valid_base_state()
{
    BaseState x;
    x << 0.0, 0.0, 0.0,
         1.0, 0.0, 0.0, 0.0;
    return x;
}

static Vec3 valid_magnetic_field()
{
    Vec3 B;
    B << 0.0, 1.0e-5, 0.0;
    return B;
}

TEST_CASE("MTQ constructor", "[mtq]")
{
    Vec3 axis = valid_axis();
    double max_dipole = 0.2;

    REQUIRE_NOTHROW(MTQ(axis, max_dipole));

    MTQ mtq(axis, max_dipole);
    REQUIRE(mtq.axis().isApprox(axis));
    REQUIRE(mtq.u_max() == max_dipole);
}

TEST_CASE("MTQ torque computation", "[mtq]")
{
    Vec3 axis;
    axis << 1.0, 0.0, 0.0;
    MTQ mtq(axis, 0.2);

    BaseState x = valid_base_state();
    Vec3 B_body;
    B_body << 0.0, 1.0e-5, 0.0;
    double u = 0.1;

    Vec3 tau = mtq.torque(u, x, B_body);

    Vec3 expected;
    expected << 0.0, 0.0, 1.0e-6;
    
    REQUIRE(tau.isApprox(expected, 1e-12));
}

TEST_CASE("MTQ torque with different configurations", "[mtq]")
{
    Vec3 axis;
    axis << 0.0, 1.0, 0.0;
    MTQ mtq(axis, 1.0);

    BaseState x = valid_base_state();
    
    Vec3 B;
    B << 2.0e-5, 0.0, 0.0;
    
    Vec3 tau = mtq.torque(0.5, x, B);
    
    Vec3 expected;
    expected << 0.0, 0.0, -1.0e-5;
    
    REQUIRE(tau.isApprox(expected, 1e-12));
}

TEST_CASE("MTQ torque is zero when B parallel to axis", "[mtq]")
{
    Vec3 axis;
    axis << 1.0, 0.0, 0.0;
    MTQ mtq(axis, 1.0);

    BaseState x = valid_base_state();
    Vec3 B;
    B << 1.0e-5, 0.0, 0.0;

    Vec3 tau = mtq.torque(1.0, x, B);
    
    REQUIRE(tau.norm() < 1e-15);
}

TEST_CASE("MTQ torque is zero when u is zero", "[mtq]")
{
    Vec3 axis = valid_axis();
    MTQ mtq(axis, 1.0);

    BaseState x = valid_base_state();
    Vec3 B = valid_magnetic_field();

    Vec3 tau = mtq.torque(0.0, x, B);
    
    REQUIRE(tau.isZero());
}

TEST_CASE("MTQ dtorq_du", "[mtq]")
{
    Vec3 axis;
    axis << 1.0, 0.0, 0.0;
    MTQ mtq(axis, 1.0);

    BaseState x = valid_base_state();
    Vec3 B;
    B << 0.0, 1.0e-5, 0.0;

    auto J = mtq.dtorq_du(0.5, x, B);

    Vec3 B_cross_axis = B.cross(axis);
    Eigen::RowVector3d expected = -B_cross_axis.transpose();
    
    REQUIRE(J.isApprox(expected, 1e-12));
}

TEST_CASE("MTQ dtorq_dbasestate with zero dB_dq", "[mtq]")
{
    Vec3 axis = valid_axis();
    MTQ mtq(axis, 1.0);

    BaseState x = valid_base_state();
    Vec3 B = valid_magnetic_field();
    Eigen::Matrix<double, 4, 3> dB_dq = Eigen::Matrix<double, 4, 3>::Zero();

    auto J = mtq.dtorq_dbasestate(0.5, x, B, dB_dq);

    REQUIRE(J.isZero());
}

TEST_CASE("MTQ ddtorq_dudu is zero", "[mtq]")
{
    Vec3 axis = valid_axis();
    MTQ mtq(axis, 1.0);

    BaseState x = valid_base_state();
    Vec3 B = valid_magnetic_field();

    auto H = mtq.ddtorq_dudu(0.5, x, B);

    REQUIRE(H.isZero());
}

TEST_CASE("MTQ ddtorq_dudbasestate with non-zero dB_dq", "[mtq]")
{
    Vec3 axis = valid_axis();
    MTQ mtq(axis, 1.0);

    BaseState x = valid_base_state();
    Vec3 B = valid_magnetic_field();
    
    Eigen::Matrix<double, 4, 3> dB_dq;
    dB_dq.setOnes();

    auto H = mtq.ddtorq_dudbasestate(0.5, x, B, dB_dq);

    REQUIRE(H.slice(0).rows() == 1);
    REQUIRE(H.slice(0).cols() == 7);
}
